Move to JonasAlfredsson/docker-nginx-certbot

docker-nginx-certbot, a Docker image made by @staticfloat , has served us well, allowing us to take advantage of Let's Encrypt certificates without investing too much effort into infrastructure. However, the constraint of having one nginx conf file per FQDN has proven to be tedious as we would often have to specify multiple files per logical group of domains that we wish to redirect.

A fork of docker-nginx-certbot by @JonasAlfredsson allows us to specify multiple FQDNs for one cert/redirection config, which might address this need.

This issue hence covers two tasks:
- [x] Move to the newer fork, and;
- [x] Rewrite the existing .conf in `/letsencrypt` to group FQDNs together

Just leaving some logs that indicate that even with Jonas' Docker distribution, Elastic Beanstalk still complains of SIGHUP being sent. I believe `nginx -s reload` uses the signal behind the scenes.
```
2021/03/01 02:27:25.306759 [ERROR] An error occurred during execution of command [app-deploy] - [PreBuildEbExtension]. Stop running the command. Error: EbExtension build failed. Please refer to /var/log/cfn-init.log for more details. 

2021/03/01 02:27:25.307185 [INFO] Executing cleanup logic
2021/03/01 02:27:25.307304 [INFO] CommandService Response: {"status":"FAILURE","api_version":"1.0","results":[{"status":"FAILURE","msg":"Engine execution has encountered an error.","returncode":1,"events":[{"msg":"Instance deployment failed. For details, see 'eb-engine.log'.","timestamp":1614565645,"severity":"ERROR"}]}]}

2021/03/01 02:27:25.307523 [INFO] Platform Engine finished execution on command: app-deploy

2021/03/01 02:30:22.545534 [INFO] Starting...
2021/03/01 02:30:22.545590 [INFO] Starting EBPlatform-PlatformEngine
2021/03/01 02:30:22.545609 [INFO] reading event message file
2021/03/01 02:30:22.545730 [INFO] no eb envtier info file found, skip loading env tier info.
2021/03/01 02:30:22.545815 [INFO] Engine received EB command cfn-hup-exec
```

Hmm, that is interesting.
However, if your solution, of only waiting 10 seconds after start of container, solves this issue I would probably just create this file:

#!/bin/bash

# Solves issue with Elastic Beanstalk failing on deploy

# https://github.com/isomerpages/isomer-redirection/issues/133

echo "Sleeping 10 seconds"

sleep 10
echo "Continuing with startup"

And then place it here in the container: /[docker-entrypoint.d](https://github.com/nginxinc/docker-nginx/tree/master/entrypoint)/00-sleeper.sh
This would probably be the easiest fix (so you don't have to overwrite any other files).

//Jonas

‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐
On Monday, March 1, 2021 2:37 AM, LoneRifle <notifications@github.com> wrote:

> Just leaving some logs that indicate that even with Jonas' Docker distribution, Elastic Beanstalk still complains of SIGHUP being sent. I believe nginx -s reload uses the signal behind the scenes.
>
> 2021/03/01 02:27:25.306759 [ERROR] An error occurred during execution of command [app-deploy] - [PreBuildEbExtension]. Stop running the command. Error: EbExtension build failed. Please refer to /var/log/cfn-init.log for more details.
>
> 2021/03/01 02:27:25.307185 [INFO] Executing cleanup logic
> 2021/03/01 02:27:25.307304 [INFO] CommandService Response: {"status":"FAILURE","api_version":"1.0","results":[{"status":"FAILURE","msg":"Engine execution has encountered an error.","returncode":1,"events":[{"msg":"Instance deployment failed. For details, see 'eb-engine.log'.","timestamp":1614565645,"severity":"ERROR"}]}]}
>
> 2021/03/01 02:27:25.307523 [INFO] Platform Engine finished execution on command: app-deploy
>
> 2021/03/01 02:30:22.545534 [INFO] Starting...
> 2021/03/01 02:30:22.545590 [INFO] Starting EBPlatform-PlatformEngine
> 2021/03/01 02:30:22.545609 [INFO] reading event message file
> 2021/03/01 02:30:22.545730 [INFO] no eb envtier info file found, skip loading env tier info.
> 2021/03/01 02:30:22.545815 [INFO] Engine received EB command cfn-hup-exec
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, [view it on GitHub](https://github.com/isomerpages/isomer-redirection/issues/135#issuecomment-787594560), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AGMYKJUQSS4PKZKCOQCLSADTBL4XHANCNFSM4YH24AOQ).

--- a/LETSENCRYPT.md
+++ b/LETSENCRYPT.md
@@ -36,12 +36,12 @@ server makes sense.
 
 ### Implementation Overview
 
-Eliot Saba (@staticfloat) maintains a Docker image based on nginx,
-which incorporates certbot. At runtime, the image runs:
+Jonas Alfredsson (@JonasAlfredsson) maintains a Docker image originally written
+by Eliot Saba (@staticfloat) which incorporates certbot into the standard nginx 
+image. At runtime, the image runs:
 
-- a bootstrap script that inspects and disables/enables config at 
-  `/etc/nginx/user.conf.d/` and `/etc/nginx/conf.d/` if 
-  they reference missing SSL files before enabling nginx, and;
+- a bootstrap script that inspects and disables/enables config at `/etc/nginx/conf.d/`
+  if they reference missing SSL files before enabling nginx, and;
 
 - a long-running while loop that does the following both at the start 
   as well once every week:
@@ -51,18 +51,12 @@ which incorporates certbot. At runtime, the image runs:
     `/etc/letsencrypt/live/<domain.gov.sg>/privkey.pem` and the file
     is either missing or expired, and;
 
-  - enable the config once the certificates are obtained by reloading
-    nginx, through the use of `kill -HUP`.
+  - enable the config once the certificates are obtained by reloading nginx.
 
-Elastic Beanstalk is sensitive to `kill -HUP` being sent to processes 
-on initialization, terminating with failure if it detects this happening.
-To mitigate this, we override the entrypoint script so that it sleeps for 
-ten seconds before spawning a poller script, consisting wholly of the 
-aforementioned while loop, so that Elastic Beanstalk is satisfied is 
-initialization is complete.
-
-See [staticfloat/docker-nginx-certbot#30](https://github.com/staticfloat/docker-nginx-certbot/issues/30) 
-for details of this quirk.
+Elastic Beanstalk does not have its CloudWatch logger immediately enabled at 
+runtime, so to ensure we have everything logged into CloudWatch, we introduce
+a script into [`/docker-entrypoint.d/`](https://github.com/nginxinc/docker-nginx/tree/master/entrypoint) 
+that makes nginx sleep for ten seconds
 
 The contact e-mail for these certificates is configured by the env var
 `CERTBOT_EMAIL`.
@@ -81,4 +75,4 @@ to `https_www_redirects.conf`
 ### Further Reading
 
 Further information can be found at the relevant GitHub 
-[repository](https://github.com/staticfloat/docker-nginx-certbot).
+[repository](https://github.com/JonasAlfredsson/docker-nginx-certbot).

--- a/README.md
+++ b/README.md
@@ -41,6 +41,20 @@ server {
 
 ```
 
+If desired, `server_name` may have multiple entries, eg, for agency.gov.sg:
+
+```
+server {
+    listen          443 ssl http2;
+    listen          [::]:443 ssl http2;
+    server_name     agency.gov.sg agency.sg www.agency.sg;
+    ssl_certificate /etc/letsencrypt/live/agency.gov.sg/fullchain.pem;
+    ssl_certificate_key     /etc/letsencrypt/live/agency.gov.sg/privkey.pem;
+    return          301 https://www.agency.gov.sg$request_uri;
+}
+
+```
+
 The files referred to in `ssl_certificate` and `ssl_certificate_key` will not exist; these will
 be created by [certbot](https://certbot.eff.org).
 

--- a/letsencrypt/jobsgohere.gov.sg.conf
+++ b/letsencrypt/jobsgohere.gov.sg.conf
@@ -1,7 +1,7 @@
 server {
     listen          443 ssl http2;
     listen          [::]:443 ssl http2;
-    server_name     jobsgohere.gov.sg;
+    server_name     jobsgohere.gov.sg www.jobsgohere.gov.sg;
     ssl_certificate /etc/letsencrypt/live/jobsgohere.gov.sg/fullchain.pem;
     ssl_certificate_key     /etc/letsencrypt/live/jobsgohere.gov.sg/privkey.pem;
     return          301 https://www.gov.sg/features/jobs-go-here;

--- a/letsencrypt/scamshield.org.sg.conf
+++ b/letsencrypt/scamshield.org.sg.conf
@@ -1,7 +1,7 @@
 server {
     listen          443 ssl http2;
     listen          [::]:443 ssl http2;
-    server_name     scamshield.org.sg;
+    server_name     scamshield.org.sg scamshield.sg www.scamshield.sg;
     ssl_certificate /etc/letsencrypt/live/scamshield.org.sg/fullchain.pem;
     ssl_certificate_key     /etc/letsencrypt/live/scamshield.org.sg/privkey.pem;
     return          301 https://www.scamshield.org.sg$request_uri;

--- a/letsencrypt/scamshield.sg.conf
+++ b/letsencrypt/scamshield.sg.conf
@@ -1,8 +1,0 @@
-server {
-    listen          443 ssl http2;
-    listen          [::]:443 ssl http2;
-    server_name     scamshield.sg;
-    ssl_certificate /etc/letsencrypt/live/scamshield.sg/fullchain.pem;
-    ssl_certificate_key     /etc/letsencrypt/live/scamshield.sg/privkey.pem;
-    return          301 https://www.scamshield.org.sg$request_uri;
-}

--- a/letsencrypt/www.jobsgohere.gov.sg.conf
+++ b/letsencrypt/www.jobsgohere.gov.sg.conf
@@ -1,8 +1,0 @@
-server {
-    listen          443 ssl http2;
-    listen          [::]:443 ssl http2;
-    server_name     www.jobsgohere.gov.sg;
-    ssl_certificate /etc/letsencrypt/live/www.jobsgohere.gov.sg/fullchain.pem;
-    ssl_certificate_key     /etc/letsencrypt/live/www.jobsgohere.gov.sg/privkey.pem;
-    return          301 https://www.gov.sg/features/jobs-go-here;
-}

--- a/letsencrypt/www.scamshield.sg.conf
+++ b/letsencrypt/www.scamshield.sg.conf
@@ -1,8 +1,0 @@
-server {
-    listen          443 ssl http2;
-    listen          [::]:443 ssl http2;
-    server_name     www.scamshield.sg;
-    ssl_certificate /etc/letsencrypt/live/www.scamshield.sg/fullchain.pem;
-    ssl_certificate_key     /etc/letsencrypt/live/www.scamshield.sg/privkey.pem;
-    return          301 https://www.scamshield.org.sg$request_uri;
-}

